compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -560,23 +566,25 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                 let def_id = did.expect_local();
                 self.visit_inner_body(def_id);
             }
-            ExprKind::Field { lhs, .. } => {
+            ExprKind::Field { lhs, variant_index, name } => {
                 let lhs = &self.thir[lhs];
-                if let ty::Adt(adt_def, _) = lhs.ty.kind()
-                    && adt_def.is_union()
-                {
-                    if let Some(assigned_ty) = self.assignment_info {
-                        if assigned_ty.needs_drop(self.tcx, self.param_env) {
-                            // This would be unsafe, but should be outright impossible since we
-                            // reject such unions.
-                            assert!(
-                                self.tcx.dcx().has_errors().is_some(),
-                                "union fields that need dropping should be impossible: \
+                if let ty::Adt(adt_def, _) = lhs.ty.kind() {
+                    if adt_def.variant(variant_index).fields[name].safety == Safety::Unsafe {
+                        self.requires_unsafe(expr.span, UseOfUnsafeField);
+                    } else if adt_def.is_union() {
+                        if let Some(assigned_ty) = self.assignment_info {
+                            if assigned_ty.needs_drop(self.tcx, self.param_env) {
+                                // This would be unsafe, but should be outright impossible since we
+                                // reject such unions.
```

Comment:
A similar "unsafe to drop" dynamic that applies to unions applies to all ADTs with unsafe fields, since an unsafe field with both:
1. a non-trivial drop
2. suspended invariant

may not be sound to drop, as its `Drop` impl may depend on the value being in a valid state.

Unions (whose fields are also unsafe to read) solve this problem by requiring that their fields are are either be `Copy` or `ManuallyDrop`, both of which preclude non-trivial drops. We should apply this same trait-based approach to unsafe fields, plus have this extra redundant check that unions get. 


See here for how the union drop check is implemented: https://github.com/rust-lang/rust/blob/d4822c2d84c242cc7403118b50c571464f38ef8f/compiler/rustc_hir_analysis/src/check/check.rs#L85-L150

Reply:
I believe I did this.